src/cargo/ops/resolve.rs

```diff
@@ -1,30 +1,101 @@
+/// taking into account `paths` overrides and activated features.
+pub fn resolve_ws_precisely<'a>(ws: &Workspace<'a>,
+                                source: Option<Box<Source + 'a>>,
+                                features: &[String],
+                                all_features: bool,
+                                no_default_features: bool,
+                                specs: &[PackageIdSpec])
+                                -> CargoResult<(PackageSet<'a>, Resolve)> {
+    let features = features.iter().flat_map(|s| {
+        s.split_whitespace()
+    }).map(|s| s.to_string()).collect::<Vec<String>>();
+
+    let mut registry = PackageRegistry::new(ws.config())?;
+
+    let mut write_lockfile = true;
+    if let Some(source) = source {
+        if let Some(root_package) = ws.current_opt() {
+            // Avoid writing a lockfile if we are `cargo install`ing a non local package.
+            write_lockfile = false;
+            registry.add_preloaded(root_package.package_id().source_id(), source);
+        }
+    }
+
+    // First, resolve the root_package's *listed* dependencies, as well as
+    // downloading and updating all remotes and such.
+    let resolve = resolve_with_registry(ws, &mut registry, write_lockfile)?;
+
+    // Second, resolve with precisely what we're doing. Filter out
+    // transitive dependencies if necessary, specify features, handle
+    // overrides, etc.
+    let _p = profile::start("resolving w/ overrides...");
+
+    add_overrides(&mut registry, ws)?;
+
+    let method = if all_features {
+        Method::Everything
+    } else {
+        Method::Required {
+            dev_deps: true, // TODO: remove this option?
+            features: &features,
+            uses_default_features: !no_default_features,
+        }
+    };
+
+    let resolved_with_overrides =
+    ops::resolve_with_previous(&mut registry, ws,
+                               method, Some(&resolve), None,
+                               &specs)?;
+
+    for &(ref replace_spec, _) in ws.root_replace() {
+        if !resolved_with_overrides.replacements().keys().any(|r| replace_spec.matches(r)) {
+            ws.config().shell().warn(
+                format!("package replacement is not used: {}", replace_spec)
+            )?
+        }
+    }
+
+    let packages = get_resolved_packages(&resolved_with_overrides, registry);
+
+    Ok((packages, resolved_with_overrides))
+}
+
+fn resolve_with_registry(ws: &Workspace, registry: &mut PackageRegistry, write_lockfile: bool)
+                         -> CargoResult<Resolve> {
     let prev = ops::load_pkg_lockfile(ws)?;
     let resolve = resolve_with_previous(registry, ws,
                                         Method::Everything,
                                         prev.as_ref(), None, &[])?;
 
-    // Avoid writing a lockfile if we are `cargo install`ing a non local package.
-    if ws.current_opt().map(|pkg| pkg.package_id().source_id().is_path()).unwrap_or(true) {
```

Comment:
This is the only actual change in the code that is in this PR. Instead of this scary condition, we basically using `source.is_some()` [other there](https://github.com/rust-lang/cargo/pull/3413/files#diff-e3d98804dece93349c90df7540234488R41). I believe these are equivalent.